app/controllers/application_controller.rb

```diff
@@ -1436,53 +1436,57 @@ def show_index_of_objects(query, args = {})
     query_params_set(query)
 
     # Supply default error message to display if no results found.
-    if (query.params.keys - query.required_parameters - [:by]).empty?
```

Comment:
Currently this PR essentially disables this condition.  I don't see a reason for it.  It only appears to disable the code from providing a more detailed error report is the key in question if there are no non-required parameters.  I tried turning it off in this PR and had to fix two tests and got a whole lot more coverage.

Reply:
I'm trying to puzzle this out, too.  The condition is essentially asking whether the user has customized the search in some way.  That is, added additional filter parameters which are not required by the flavor.  There are a whole slew of these filters one can apply to all observation queries, for example.  You can restrict the query to a given user, certain dates, months of the year, names or clades (Agaricales or lichens, say), confident observations, ones with sequences, etc. etc. etc.

If someone did this, for example, to a "by user" query -- let's say, give me all of Alan's observations restricted to Antarctica.  This is probably empty (although who knows, given it's Alan! :).  This code will produce the error message "There are no observations by Alan".  That seems misleading.  The default error message "there are no matching observations" seems more appropriate.

Of course, I don't know how someone would actually effect such a query.  You can easily add filters like "region:Antarctica" to pattern searches.  But the error message for pattern searches is "no observations matching "blah blah blah"", which seems entirely appropriate.  You can add site-wide content filters via the user prefs.  This is currently restricted to: 1) only show obs in a given clade, 2) only show obs with images, 3) only show obs with specimens, 4) only show obs from a given region.

So here's the question, I guess: What should the error message be when someone asks to see, e.g., all of a give user's observations, but has a content filter which yields an empty result?  That is, say, if you click on "Your Observations" -- "There are no observations by Nathan"?  "There are no matching observations?"  Or... ?

I don't know what the ideal answer is.  But I think this is a fair summary of the context, so that you can make an informed decision about whether to change the behavior or not while refactoring this.

Enjoy! :)

Reply:
Thanks!  I got buried in a semi-urgent disk recovery situation that has been a lot more painful than expected.

Anyway, the challenge I have with the current code is there doesn't appear to be an obvious way to get to the more specific error messages as the code is currently written, but if you turn off that condition you can.  Personally, I don't see a big problem with "There are no observations by Alan" if the person has somehow bent over backwards trying to get observations about Antarctica by Alan, then they should have the context.  However, I'm also fine with the more generic error.  That's certainly the simplest thing to do and it eliminates code which is great for helping with the ABC metric!  Given your descriptions, I'm leaning towards just making it simple and having a single error message.